src/dos/dos_programs.cpp

```diff
@@ -129,37 +130,39 @@ class MOUNT : public Program {
 			if (DOS_GetDefaultDrive() == 25) DOS_SetDrive(i_newz);
 		}
 	}
-	void ListMounts(void) {
-		char name[DOS_NAMELENGTH_ASCII];Bit32u size;Bit16u date;Bit16u time;Bit8u attr;
-		/* Command uses dta so set it to our internal dta */
-		RealPt save_dta = dos.dta();
-		dos.dta(dos.tables.tempdta);
-		DOS_DTA dta(dos.dta());
+
+	void ListMounts()
+	{
+		const std::string header_drive = MSG_Get("PROGRAM_MOUNT_STATUS_DRIVE");
+		const std::string header_type = MSG_Get("PROGRAM_MOUNT_STATUS_TYPE");
+		const std::string header_label = MSG_Get("PROGRAM_MOUNT_STATUS_LABEL");
+
+		const int term_width = real_readw(BIOSMEM_SEG, BIOSMEM_NB_COLS);
+		const auto width_1 = static_cast<int>(header_drive.size());
+		const auto width_3 = std::max(11, static_cast<int>(header_label.size()));
+		const auto width_2 = term_width - 3 - width_1 - width_3;
+
+		auto print_row = [&](const std::string &txt_1,
+		                     const std::string &txt_2,
+		                     const std::string &txt_3) {
+			WriteOut("%-*s %-*s %-*s\n",
+			         width_1, txt_1.c_str(),
+			         width_2, txt_2.c_str(),
+			         width_3, txt_3.c_str());
+		};
 
 		WriteOut(MSG_Get("PROGRAM_MOUNT_STATUS_1"));
-		WriteOut(MSG_Get("PROGRAM_MOUNT_STATUS_FORMAT"),
-		         MSG_Get("PROGRAM_MOUNT_STATUS_DRIVE"),
-		         MSG_Get("PROGRAM_MOUNT_STATUS_TYPE"),
-		         MSG_Get("PROGRAM_MOUNT_STATUS_LABEL"));
-		for (int p = 0;p < 8;p++) WriteOut("----------");
-
-		for (int d = 0; d < DOS_DRIVES; d++) {
-			if (!Drives[d]) continue;
-
-			char root[7] = {static_cast<char>('A' + d),':','\\','*','.','*',0};
-			bool ret = DOS_FindFirst(root,DOS_ATTR_VOLUME);
-			if (ret) {
-				dta.GetResult(name,size,date,time,attr);
-				DOS_FindNext(); //Mark entry as invalid
-			} else {
-				name[0] = 0;
+		print_row(header_drive, header_type, header_label);
+		for (int i = 0; i < term_width; i++)
+			WriteOut_NoParsing("-");
+
+		for (uint8_t d = 0; d < DOS_DRIVES; d++) {
+			if (Drives[d]) {
+				print_row(std::string{drive_letter(d)},
+				          Drives[d]->GetInfo(),
+				          To_Label(Drives[d]->GetLabel()));
```

Comment:
:+1: This is a lot more understandable!

Reply:
The previous implementation seemed like a hack that was implemented very early - perhaps there was no functionality to obtain label directly from disk object back then…

I think the same hack (start directory listing and finish it immediately) might be in use in other places in code.